### ISSUE TITLE
generate-dhparam becomes a zombie process

### ISSUE THREAD

**Issue by frederikbosch** [jwilder/nginx-proxy#1161]
When running a nginx proxy the generation of a dhparam file is executed in the background. However, this causes the process to become a zombie process.

```log
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
xxxx     16431  0.0  0.0      0     0 ?        Z    13:07   0:00 [generate-dhpara] <defunct>
```

My bash knowledge is limited, but I guess this is caused [in these lines](https://github.com/jwilder/nginx-proxy/blob/master/generate-dhparam.sh#L28:L45). Maybe someone with bash knowledge knows how to prevent this.

**Comment by gerrytsui** [comment]
Yep. we had the same issue, and it got worst running in the latest kernel (that comes with Ubuntu 18.04).

Please try update the last line of Dockerfile with
CMD ["/bin/sh","-c","forego start -r"]
( you will have to build from Dockerfile instead of docker pull from docker hub )

it has to do with the forego creating a "&disown" process and becoming an "zombie" when the parent process was end.

**Comment by frederikbosch** [comment]
@gerrytsui Exacty, we are running Ubuntu 18.04 too. I haven't tried your suggestion out, but why would that call make a difference compared to the current `CMD ["forego", "start", "-r"]`? Just out of interest.

**Comment by gerrytsui** [comment]
@frederikbosch The difference is (within the Docker container) /bin/sh will run as PID 1 (instead of forego). I am not an expert in Linux process (reaping) management, but I believe the go-compiler produced forego can not handle linux signals properly with the new Linux kernel. 

I was getting hundreds of zombie processes in a few hours ( every time nginx get a reload signal, from docker-gen activities), after upgraded to 18.04 back in April, and fixed it temporary with 18.04 running on top of Kernel 4.4 ( instead of 4.15)

You can  inspect the process tree of PID 1  inside the container (you will need to install procps), with  ps auxf command

Another way to fix this is to use the tini  or the docker --init, but (was/is) not available yet from our orchestration layer : docker-compose.

**Comment by gerrytsui** [comment]
@frederikbosch  OK I tested the better workaround:  we upgraded our docker environment to use docker-compose file version 3.7 and docker-engine 18.06 and are able to eliminate zombie processes by specifying init: true ( under nginx service in the yml), without modifying the CMD line. 

If you don't use docker-compose, you should be able to specify docker run --init to achieve the same result

pstree command shows that the forego is now a child of init instead of /bin/sh.

**Comment by frederikbosch** [comment]
@gerrytsui Perfect. I do. Will test it out, hopefully tomorrow. Many thanks for your help, greatly appreciated.

**Comment by saostad** [comment]
> @frederikbosch OK I tested the better workaround: we upgraded our docker environment to use docker-compose file version 3.7 and docker-engine 18.06 and are able to eliminate zombie processes by specifying init: true ( under nginx service in the yml), without modifying the CMD line.
> 
> If you don't use docker-compose, you should be able to specify docker run --init to achieve the same result
> 
> pstree command shows that the forego is now a child of init instead of /bin/sh.

it worked for me.
